python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,558 @@
+        pass
+    return DimensionNameSet(element.name if isinstance(element, DimensionElement) else element
+                            for element in container)
+
+
+class DimensionSet(Set):
+    r"""A custom set/dict hybrid class for collections of `DimensionElement`\s.
+
+    `DimensionSet` objects implement the full (immutable)
+    `collections.abc.Set` interface.  In addition, like `frozenset`, they are
+    immutable and hashable, and also provide named-method versions of most
+    operators.  Unlike Python sets, they are deterministically sorted.
+
+    `DimensionSet` comparison operators and named-method relation operations
+    accept other set-like objects and iterables containing either
+    `DimensionElement` instances or their string names; because
+    `DimensionElement`\s cannot be directly constructed, APIs that accept them
+    should generally accept a name as an alternative when the transformation
+    to a `DimensionElement` can be done internally.  Operators that return
+    new sets (`|`, `&`, `^`, and `^`) do require `DimensionSet` operands on
+    both sides to avoid surprises in return types.
+
+    Because the `DimensionElement` objects they hold always have a name,
+    `DimensionSet`\s also supports some `dict`-like operations: including
+    regular square-bracket indexing (`__getitem__`), `get`, and the ``in``
+    operator (`__contains__`).  Both names and `DimensionElement` objects can
+    be passed to any of these. The `names` attribute can also be used to
+    obtain a `set`-like object containing those names.
+
+    `DimensionSet` instances cannot be constructed directly; they can only be
+    obtained (possibly indirectly) from a special "universe" `DimensionGraph`
+    loaded from configuration.
+
+    Parameters
+    ----------
+    univsere : `DimensionGraph`
+        Ultimate-parent `DimensionGraph` that constructed the elements in
+        this set.
+    elements : iterable of `DimensionElement` or `str`
+        Elements to include in the set, or names thereof.
+    expand : `bool`
+        If `True`, recursively expand the set to include dependencies.
+    optional : `bool`
+        If `True`, include optional dependencies in expansion.  Ignored
+        if ``expand`` is `False`.
+
+    """
+    def __init__(self, universe, elements, expand=False, optional=False):
+        self._universe = universe
+        self._elements = OrderedDict()
+
+        def toPairs(elems):
+            for elem in elems:
+                if not isinstance(elem, DimensionElement):
+                    elem = self._universe.elements[elem]
+                yield (elem.name, elem)
+                if expand:
+                    yield from toPairs(elem.dependencies(optional=optional))
+
+        names = dict(toPairs(elements))
+        if names:
+            for candidate in self._universe.elements:
+                if names.pop(candidate.name, None):
+                    self._elements[candidate.name] = candidate
```

Comment:
Yes, it's all about the insertion order, and of course this is only a reasonable approach to that because we know the set of Dimensions is going to be small and hence making everything (at least) O(N) in the number of dimensions in universe is not a problem.  I'll add a comment covering all of that.